Configurable email backends

jrief opened the pull request:
# Trac ticket number

ticket-35514

# Branch description

This is a draft pull request. It just replaces the the configuration settings starting with `EMAIL_…` with a dictionary based configuration. More features have to be added.

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

medmunds commented on django/core/mail/__init__.py:
```diff
@@ -89,7 +117,12 @@ def send_mail(
+    auth_user=None,
+    auth_password=None,
+    connection=None,
 ):
```
(Same comments as above about argument order.)

medmunds replied:
[resolved]

medmunds commented on django/core/mail/__init__.py:
```diff
@@ -103,11 +136,24 @@ def send_mass_mail(
-        password=auth_password,
-        fail_silently=fail_silently,
-    )
+    if provider:
+        if connection or auth_user or auth_password:
+            raise ValueError(
+                "provider and connection/auth_user/auth_password are mutually "
+                "exclusive, so only use either of those arguments."
+            )
+        provider_settings = settings.EMAIL_PROVIDERS[provider]
+        connection = get_connection(
+            backend=provider_settings["BACKEND"],
+            fail_silently=fail_silently,
+            **provider_settings["OPTIONS"],
+        )
+    elif not connection:
+        connection = get_connection(
+            username=auth_user,
+            password=auth_password,
+            fail_silently=fail_silently,
+        )
```
(Same comments as above about re-using provider logic from get_connection().)

medmunds commented on django/core/mail/__init__.py:
```diff
@@ -116,7 +162,12 @@ def send_mass_mail(
 
 
 def mail_admins(
-    subject, message, fail_silently=False, connection=None, html_message=None
+    subject,
+    message,
+    fail_silently=False,
+    provider=None,
+    connection=None,
+    html_message=None,
```
(Same comments as above about argument order.)

medmunds commented on django/conf/global_settings.py:
```diff
@@ -184,29 +184,53 @@ def gettext_noop(s):
+            "ssl_certfile": None,
+            "ssl_keyfile": None,
+        },
+        "USE_LOCALTIME": False,
+        "TIMEOUT": None,
```
I feel like `USE_LOCALTIME` and `TIMEOUT` should both be within `OPTIONS`.

medmunds commented on django/core/mail/backends/base.py:
```diff
@@ -15,8 +15,9 @@ class BaseEmailBackend:
-    def __init__(self, fail_silently=False, **kwargs):
+    def __init__(self, fail_silently=False, use_localtime=False, **kwargs):
         self.fail_silently = fail_silently
+        self.use_localtime = use_localtime
```
I'm not thrilled about adding `use_localtime` to the BaseEmailBackend. It doesn't apply to a lot of third-party email backends.

I get that you're trying to share code between Django's built-in EmailBackends. I don't have a great solution, other than duplicating the init code to the backends that use it.

jrief replied:
Looking at the code again, I probably have to move `use_localtime` to `OPTIONS` as you suggested previously. Since the backend is responsible to generate the dates in the emails, this otherwise can not work properly.

jrief replied:
Just had another look. `use_localtime` is only used in the console backend. I therefore moved that parameter from the `BaseEmailBackend` to the `console.EmailBackend`. I believe that's the most unintrusive solution.

medmunds replied:
I think `use_localtime` would be needed for any backend that calls EmailMessage.message(), which includes Django's smtp, filebased, console and locmem. (And I guess any third-party backends that call it.)

Agreed that `use_localtime` doesn't really feel like a backend connection option.

An alternative would be to leave `EMAIL_USE_LOCALTIME` as a global setting, outside `EMAIL_PROVIDERS` (and continue using that global setting in EmailMessage.message()). That's probably the best for compatibility, but loses the ability to have different `use_localtime` settings for different providers.

medmunds commented on django/core/mail/backends/smtp.py:
```diff
@@ -18,37 +18,25 @@ class EmailBackend(BaseEmailBackend):
-        timeout=None,
-        ssl_keyfile=None,
-        ssl_certfile=None,
-        **kwargs,
+        use_localtime=False,
+        **options,
```
I think it would be better to keep the individual args listed, rather than collapsing them all into `**options`. Individual args are clearer for other people reading the code: e.g., to find out what they can include in settings OPTIONS or pass as overrides to get_connection(). (And `**options` complicates efforts like django-stubs.)

medmunds commented on django/core/mail/backends/smtp.py:
```diff
@@ -18,37 +18,25 @@ class EmailBackend(BaseEmailBackend):
-            settings.EMAIL_SSL_CERTFILE if ssl_certfile is None else ssl_certfile
-        )
+        super().__init__(fail_silently=fail_silently, use_localtime=use_localtime)
+        self.host = options.pop("host", None)
+        self.port = options.pop("port", None)
+        self.username = options.pop("username", None)
+        self.password = options.pop("password", None)
+        self.use_tls = options.pop("use_tls", None)
+        self.use_ssl = options.pop("use_ssl", None)
+        self.ssl_keyfile = options.pop("ssl_keyfile", None)
+        self.ssl_certfile = options.pop("ssl_certfile", None)
```
I may be missing something, but don't you still need to pull in the deprecated settings values here if this is *not* being initialized from an EMAIL_PROVIDERS OPTIONS dict?

(Otherwise that logic has to go in mail.get_connection(), as a special case for the SMTP EmailBackend, which seems complicated.)

Also, I think this needs to know whether it's being initialized from a provider or not. If it's from a provider, OPTIONS dict, it shouldn't use *any* deprecated settings. If it's *not* from a provider, deprecated settings are OK if not provided as args. (But you don't want to mix the two.) I'll add a separate comment about that at the PR level.

jrief replied:
This is true. I actually could add something here: https://github.com/jrief/django/blob/feae9bc4b4836175089365d0b5d0bfe3b21735f7/django/conf/__init__.py#L236

```python
self.EMAIL_PROVIDER["default"]["BACKEND"] = self.EMAIL_BACKEND
```

The problem then however is how to handle configurations which define the old and the new settings?

Since configurations are handled by the user of a library, for them it would be easy to rewrite them before deploying a new major version of Django.

medmunds commented on django/core/mail/message.py:
```diff
@@ -209,6 +209,7 @@ def __init__(
         from_email=None,
         to=None,
         bcc=None,
+        provider=None,
```
1. Shouldn't change existing arg order; suggest kwarg only.
2. But: I think the `provider` arg actually belongs in EmailMessage.send(), not in the constructor.

I think it was a mistake to include `connection` as an EmailMessage constructor arg. It really only applies when it's time to send the message. To demonstrate the problem, which server will be used to send these messages?

```python
from django.core.mail import get_connection, EmailMessage

conn1 = get_connection(host="smtp.gmail.com", ...)
msg1 = EmailMessage(connection=conn1, ...)
msg2 = EmailMessage(connection=conn1, ...)

conn2 = get_connection(host="smtp.example.org", ...)
conn2.send_messages([msg1, msg2])
```

We can't easily fix this for with EmailMessage.connection right now, but we should avoid extending it to the new `provider` option.

jrief replied:
ad 1) fixed
ad 2) as long as `connection` is part of the constructor, `provider` is the expected successor.

medmunds replied:
Well, I don't think `connection` belongs in the constructor either. (It should have been an arg to EmailMessage.send().) `provider` could still be the successor, just moved to the method where it applies.

If we really want to add `provider` to the constructor, I think we should document that it will be ignored when [using `connection.send_messages()` as described in the docs](https://docs.djangoproject.com/en/5.0/topics/email/#sending-multiple-emails).

medmunds reviewed (changes requested):
Hooray! Nice work—this is seriously very exciting.

medmunds commented:
I *think* an EmailBackend's init is going to need to know whether or not it was initialized from an EMAIL_PROVIDERS OPTIONS dict, to decide whether it's OK to use deprecated settings. See [this review comment](#discussion_r1695981045).

My suggestion would be that get_connection() should pass a new `provider_alias` (or `_provider`) param to the init:

```patch
  def get_connection(backend=None, fail_silently=False, *, provider=None, **kwds):
+     provider_alias = provider
      if backend:
          ...
-         return klass(fail_silently=fail_silently, **kwds)
          return klass(
               fail_silently=fail_silently, 
+              provider_alias=provider_alias, 
               **kwds,
           )
      provider = provider or "default"
      klass = import_string(settings.EMAIL_PROVIDERS[provider]["BACKEND"])
      return klass(
          fail_silently=fail_silently,
+         provider_alias=provider_alias,
          **(settings.EMAIL_PROVIDERS[provider]["OPTIONS"] | **kwds),
      )
```

Then EmailBackend init methods can check `if provider_alias is None` before using deprecated `EMAIL_*` settings.

(The new arg could just be called `provider`, but I think that name has a relatively high chance of conflicting with some existing EmailBackend's init args.)

medmunds commented:
(Also, I think this is missing tests that the deprecated settings still work for now.)

jrief commented on django/core/mail/__init__.py:
```diff
@@ -40,16 +40,30 @@
+    klass = import_string(settings.EMAIL_PROVIDERS[provider]["BACKEND"])
+    return klass(
+        fail_silently=fail_silently,
+        use_localtime=settings.EMAIL_PROVIDERS[provider]["USE_LOCALTIME"],
```
Because in my opinion this should not be part of the connection settings. Instead it should rather be part of the email generation process. We might further discuss this.

jrief commented:
Many thanks to your suggestions, after reading them I believe that it makes sense to disallow the deprecated `EMAIL_…` settings. Rewriting them in the owner's `settings.py` is easy a job we can expect from users migrating to Django-6.0. After all, it's a major release and hence breaking changes may be expected. In addition to that, the user is well informed and knows what to to after upgrading to Django-6.0.

Having a compatibility layer makes the code really complicated for a minimal payoff. I will propose this problem to the community chat on Discord and ask for advise there.

medmunds commented:
@jrief thanks for the followup. I added some more info about the deprecated settings and compatibility, but agree a forum discussion would be easier.

There are a handful of other comments you may not have seen. In particular, I think any new args to `send_mail()` and other documented functions have to be added at the end of the list, for compatibility and because ["The API for this method is frozen"](https://github.com/django/django/pull/18421/files#diff-a4a410070ddd9a2614dbf98311b960fbe7dd6fcacb2b9ace85dac1caa50c2b1aL74). (It's unfortunate they can't go where they'd make logical sense, which is why I'm suggesting keyword-only args going forward.)

(GitHub has an annoying habit of hiding review comments behind a "load more" link. Search for "hidden conversations" in the thread above.)

dpgraham4401 commented:
Hi @jrief, do you still have to bandwidth to work on this? I'm happy to pitch in.

jrief commented:
@dpgraham4401  sorry for not proceeding on this. Currently there is one missing unit test and I haven't found any viable way on how to fix it. Shall we chat on this?

dpgraham4401 commented:
@jrief I'll take a look after the holidays. Happy to chat as well

jrief commented:
@dpgraham4401 I had some time to work on this feature. I implemented the remaining unit tests and rebased to HEAD of the main branch.

@medmunds This pull request now is ready for another review. What I did, is not moving the settings `EMAIL_USE_LOCALTIME` to the new configuration directives. This means that you can not configure this, depending on the used backend. This is caused by the API freeze mentioned before.

medmunds commented:
@jrief thanks! I'll try to take a closer look over the weekend.

A few quick notes:

* Keeping `EMAIL_USE_LOCALTIME` as a separate setting (outside `EMAIL_PROVIDERS`) seems fine to me. (It looks like there's some leftover code from your earlier approach in get_connection().)
* There are still a handful of unresolved comments from my initial review: search for "hidden conversations" on this page, or look in the "files changed" tab. In particular, the order of existing arguments to send_message() and other functional APIs needs to be preserved to avoid breaking existing code. (I wish there were some way we could deprecate posargs to those functions, to simplify changes like this one. Maybe that's a separate ticket.)
* The deprecations will all need to be changed from RemovedInDjango61Warning to RemovedInDjango70Warning. (Django 5.2 just entered feature freeze, so this PR is now targeting Django 6.0, and deprecations live for at least two releases.)
* Do we have a solution to the "wrapper backend" compatibility problem from [my earlier comment](https://github.com/django/django/pull/18421#discussion_r1700878228)?

jrief commented:
I also moved `EMAIL_FILE_PATH` into `EMAIL_PROVIDERS[…]['OPTIONS']['file_path']` to be consistent with the other backends. It currently is only used in the `django.core.mail.backends.filebased.EmailBackend`.

pauloxnet commented:
If this PR isn't a draft anymore, maybe you can remove WIP from the title to avoid confusion.

medmunds commented:
@jrief part of Django's deprecation policy is that existing code continues to work (with warnings) during the deprecation period. So until a user chooses to change their settings.py to EMAIL_PROVIDERS (or Django 7.0 is released), the existing EMAIL_* settings will need to work.

Here are a few examples I would expect to work (and they don't):
```python
@ignore_warnings(category=RemovedInDjango70Warning)
class ExampleDeprecatedSettingsTests(SimpleTestCase):
    # During deprecation period, existing EMAIL_* settings must continue to work.
    @override_settings(
        EMAIL_BACKEND="django.core.mail.backends.smtp.EmailBackend",
        EMAIL_HOST="mail.example.com",
        EMAIL_PORT=9876,
    )
    def test_deprecated_settings_still_work(self):
        connection = mail.get_connection()
        self.assertIsInstance(connection, smtp.EmailBackend)
        self.assertEqual(connection.host, "mail.example.com")
        self.assertEqual(connection.port, 9876)

    @override_settings(
        EMAIL_BACKEND="django.core.mail.backends.smtp.EmailBackend",
        EMAIL_HOST="mail.example.com",
        EMAIL_PORT=9876,
    )
    def test_deprecated_settings_with_get_connection_kwargs(self):
        # kwarg params to get_connection() override settings
        connection = mail.get_connection(host="other.example.net")
        self.assertIsInstance(connection, smtp.EmailBackend)
        self.assertEqual(connection.host, "other.example.net")
        self.assertEqual(connection.port, 9876)  # from settings

    @override_settings(
        EMAIL_BACKEND="django.core.mail.backends.dummy.EmailBackend",
        EMAIL_HOST="mail.example.com",
        EMAIL_PORT=9876,
    )
    def test_get_connection_specific_backend(self):
        # specific backend overrides EMAIL_BACKEND setting; kwargs work as above
        connection = mail.get_connection(
            "django.core.mail.backends.smtp.EmailBackend",
            host="other.example.net",
        )
        self.assertIsInstance(connection, smtp.EmailBackend)
        self.assertEqual(connection.host, "other.example.net")
        self.assertEqual(connection.port, 9876)  # from settings
```

(Those are just examples. It would be best to keep copies of all the existing settings/configuration/get_connection() params tests in a "test deprecated features" section, to make sure everything continues to work during the deprecation period. Django's developer docs have some suggestions for updating tests when [*deprecating a feature*](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature).)

I *suspect* maintaining settings compatibility would be cleaner if each EmailBackend.\_\_init__() stays responsible for loading its own deprecated settings, rather than trying to handle this mapping in the settings manager or get_connection(). If so, that logic would probably look something like:

```python
class EmailBackend(BaseEmailBackend):  # SMTP backend
    def __init__(self, host=None, port=None, ...):
        if (being initialized from EMAIL_PROVIDERS OPTIONS dict):
            self.host = host
            self.port = port
            ...
        else:
            # RemovedInDjango70Warning.
            # Not using EMAIL_PROVIDERS, so fall back to deprecated EMAIL_* settings.
            # (This just is the original EmailBackend.__init__() logic.)
            self.host = host or settings.EMAIL_HOST
            self.port = port or settings.EMAIL_PORT
            ...
```

jrief commented:
@medmunds thanks for pointing this out. I now changed the internal logic and added some extra tests.

medmunds reviewed (changes requested):
@jrief I haven't had time to go through all of the tests or docs, but wanted to get you some other feedback.

[GitHub will probably hide some of the feedback. Search for hidden comments in the Conversation tab, or go to the Files tab in the PR.]

A few high-level thoughts:

* This is great progress! It's come a long way since the last time I went through it.

* I think it would be helpful to centralize all the logic for accessing EMAIL_PROVIDERS in get_connection()—including resolving backend classes and OPTIONS. In general, if other code needs to look at EMAIL_PROVIDERS, that may indicate a larger issue we should investigate.

* I'm a little confused by some of what you're doing with settings. (I need to spend some more time poking around in it.)

  But it occurs to me that we'll need to tell third-party EmailBackend developers how to update their code for the transition to EMAIL_PROVIDERS. This includes explaining how they should handle their own settings during the deprecation period. And we should be able to point them to Django's smtp.EmailBackend as an example.

  If the only way to get Django's SMTP backend working through the transition is by using undocumented functions like settings.is_overridden(), or by adding special case handling deep in Django's settings implementation, that concerns me. Because third-party packages can't do those things for their own settings. I'm not really sure how to handle this—it's actually what's delayed my review. I'll keep investigating.

Again, though, this is some really nice work, in an extremely complicated area. Thanks again for taking it on!

jrief commented on django/conf/global_settings.py:
```diff
@@ -186,29 +186,51 @@ def gettext_noop(s):
+EMAIL_PROVIDERS = {
+    "default": {
+        "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
+        "OPTIONS": {
+            "host": "localhost",
+            "port": 25,
+            "username": "",
+            "password": "",
+            "use_tls": False,
+            "use_ssl": False,
+            "ssl_certfile": None,
+            "ssl_keyfile": None,
+            "timeout": None,
+        },
```
sure. I removed them

jrief commented on django/conf/global_settings.py:
```diff
@@ -186,29 +186,51 @@ def gettext_noop(s):
-EMAIL_SSL_CERTFILE = None
-EMAIL_SSL_KEYFILE = None
-EMAIL_TIMEOUT = None
+dep_EMAIL_HOST_USER = ""
+dep_EMAIL_HOST_PASSWORD = ""
+dep_EMAIL_USE_TLS = False
+dep_EMAIL_USE_SSL = False
+dep_EMAIL_SSL_CERTFILE = None
+dep_EMAIL_SSL_KEYFILE = None
+dep_EMAIL_TIMEOUT = None
```
actually, when I started implementing this feature in June 24, there where other deprecated settings prefixed with `dep_…`. I just copied that pattern. They are never used, so I removed them.

medmunds replied:
[resolved]

jrief commented on django/core/mail/__init__.py:
```diff
@@ -47,13 +47,33 @@
 def get_connection(backend=None, fail_silently=False, **kwds):
```
`kwds` was not my invention, I just used it.

`provider` now is part of the params list, but this was also true in a [former version of this branch](https://github.com/jrief/django/blob/9672bca41da9b01e26bc7f25e28e2b1a9d9944f7/django/core/mail/__init__.py) .

medmunds replied:
> `kwds` was not my invention, I just used it.

Understood. Just taking the opportunity to clean up code while we're here.

> `provider` now is part of the params list, but this was also true in a [former version of this branch](https://github.com/jrief/django/blob/9672bca41da9b01e26bc7f25e28e2b1a9d9944f7/django/core/mail/__init__.py) .

I was wondering why you changed that. [My earlier comment](https://github.com/django/django/pull/18421#discussion_r1695897122) was suggesting using Python's [keyword-only parameter](https://peps.python.org/pep-3102/) syntax (with `*,` in the parameter list) for send_mail() and the other functions you were changing.

Incidentally, I've started a [forum discussion](https://forum.djangoproject.com/t/change-send-mail-and-emailmessage-to-kwargs-only/38239) about making most of the older parameters to these functions keyword-only as well. (But that will be a separate ticket/PR. For this PR, it's sufficient to add the new `provider` params as keyword-only.)

jrief commented on django/core/mail/__init__.py:
```diff
@@ -47,13 +47,33 @@
     """
-    klass = import_string(backend or settings.EMAIL_BACKEND)
-    return klass(fail_silently=fail_silently, **kwds)
+    provider = kwds.pop("provider", None)
+    if backend:
+        if provider:
+            raise ValueError(
+                "backend and provider are mutually exclusive, "
+                "so only use either of those arguments."
+            )
+        klass = import_string(backend)
+        return klass(fail_silently=fail_silently, **kwds)
+
+    if settings.is_overridden("EMAIL_BACKEND"):
+        klass = import_string(settings.EMAIL_BACKEND)
+        return klass(fail_silently=fail_silently, **kwds)
+
+    provider = provider or getattr(settings, "DEFAULT_EMAIL_PROVIDER_ALIAS", "default")
+    klass = import_string(settings.EMAIL_PROVIDERS[provider]["BACKEND"])
+    return klass(
+        fail_silently=fail_silently,
+        use_localtime=settings.EMAIL_USE_LOCALTIME,
+        **(settings.EMAIL_PROVIDERS[provider]["OPTIONS"] | kwds),
+    )
```
done

jrief commented on django/core/mail/__init__.py:
```diff
@@ -78,11 +99,24 @@ def send_mail(
-        password=auth_password,
-        fail_silently=fail_silently,
-    )
+    if provider:
+        if connection or auth_user or auth_password:
+            raise ValueError(
+                "provider and connection/auth_user/auth_password are mutually "
+                "exclusive, so only use either of those arguments."
+            )
+        provider_settings = settings.EMAIL_PROVIDERS[provider]
+        connection = get_connection(
+            backend=provider_settings["BACKEND"],
+            fail_silently=fail_silently,
+            **provider_settings["OPTIONS"],
+        )
+    else:
+        connection = connection or get_connection(
+            username=auth_user,
+            password=auth_password,
+            fail_silently=fail_silently,
+        )
```
fixed

medmunds replied:
[resolved]

jrief commented on django/core/mail/backends/smtp.py:
```diff
@@ -31,22 +31,46 @@ def __init__(
+            )
+            self.timeout = settings.EMAIL_TIMEOUT if timeout is None else timeout
+        else:
+            options = settings.EMAIL_PROVIDERS[provider or "default"]["OPTIONS"]
+            self.host = options.get("host") if host is None else host
+            self.port = options.get("port") if port is None else port
+            self.username = options.get("username") if username is None else username
+            self.password = options.get("password") if password is None else password
+            self.use_tls = options.get("use_tls") if use_tls is None else use_tls
+            self.use_ssl = options.get("use_ssl") if use_ssl is None else use_ssl
+            self.ssl_keyfile = (
+                options.get("ssl_keyfile") if ssl_keyfile is None else ssl_keyfile
+            )
+            self.ssl_certfile = (
+                options.get("ssl_certfile") if ssl_certfile is None else ssl_certfile
+            )
+            self.timeout = options.get("timeout") if timeout is None else timeout
```
fixed

jrief commented on django/core/mail/backends/smtp.py:
```diff
@@ -31,22 +31,46 @@ def __init__(
+                    "so only use either of those arguments."
+                )
+            self.host = host or settings.EMAIL_HOST
+            self.port = port or settings.EMAIL_PORT
```
fixed

jrief commented on django/core/mail/message.py:
```diff
@@ -255,12 +256,16 @@ def __init__(
                     self.attach(*attachment)
         self.extra_headers = headers or {}
         self.connection = connection
+        self.provider = provider or "default"
```
fixed

jrief commented on docs/topics/email.txt:
```diff
@@ -76,7 +88,7 @@ a secure connection is used.
 ``send_mail()``
 ===============
 
-.. function:: send_mail(subject, message, from_email, recipient_list, fail_silently=False, auth_user=None, auth_password=None, connection=None, html_message=None)
+.. function:: send_mail(subject, message, from_email, recipient_list, fail_silently=False, html_message=None, provider=None)
```
I would deprecate `connection` on the long term, because it causes more confusion than benefit. However, not in this pull request.

jrief commented on docs/topics/email.txt:
```diff
@@ -607,28 +626,44 @@ SMTP backend
+want to specify an alternative, put the following in your settings::
+
+    EMAIL_PROVIDERS = {
+        "default": {
+            "BACKEND": "django.core.mail.backends.console.EmailBackend",
+            "OPTIONS": {...},
+        },
```
ups. Fixed.

jrief commented on tests/auth_tests/test_forms.py:
```diff
@@ -1357,7 +1358,14 @@ def test_save_html_email_template_name(self):
+        EMAIL_PROVIDERS={
+            "default": {
+                "BACKEND": "mail.custombackend.FailingEmailBackend",
+                "OPTIONS": settings.EMAIL_PROVIDERS["default"]["OPTIONS"],
```
At first I focused on the smtp backend and oversaw that the other backends do not need an OPTIONS dict. Fixed.

jrief commented on tests/mail/tests.py:
```diff
@@ -1081,7 +1084,12 @@ def test_backend_arg(self):
-        EMAIL_BACKEND="django.core.mail.backends.locmem.EmailBackend",
+        EMAIL_PROVIDERS={
+            "default": dict(
+                settings.EMAIL_PROVIDERS["default"],
```
same as before: I used the SMTP backend and oversaw that the other backends do not need an OPTIONS dict. Fixed.

jrief commented:
blacken-docs fails, but I can't fix any files.

medmunds commented:
I think I figured out what's been confusing me with the settings. (Apologies if this seems long or repeats things you already know—I'm explaining it to myself, too. Scroll down to the boldface line for the conclusions.)

Before this change, the SMTP EmailBackend relied on defaults from Django's global_settings:

```python
# django/core/mail/backends/smtp.py:
class EmailBackend(BaseEmailBackend):
    def __init__(self, ..., host=None, port=None, ...):
        ...
        self.host = host or settings.EMAIL_HOST
        self.port = port or settings.EMAIL_PORT
        ...

# django/conf/global_settings.py:
EMAIL_HOST = "localhost"
EMAIL_PORT = 25
```

So if the user hadn't set `EMAIL_PORT` in their own settings.py, the SMTP EmailBackend would get its default port from Django's global defaults. It could be certain that `self.port` would never be `None`, because the global default setting would always be there as a fallback.

In implementing EMAIL_PROVIDERS, you've tried to map those old defaults into the default provider's options:

```python
# django/conf/global_settings.py:
EMAIL_PROVIDERS = {
    "default": {
        "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
        "OPTIONS": {
            "host": "localhost",
            "port": 25,
            ...
        },
    },
}
```

Which made sense to me at first: this seems like the natural replacement for the deprecated settings. It's how a user would update their own settings.py.

But the more I looked at it, I realized this is conflating two different things:
1. The default email provider (the default value of the EMAIL_PROVIDERS setting)
2. The default OPTIONS for an SMTP email backend (*any* SMTP backend)—which used to come from EMAIL_HOST, EMAIL_PORT, etc.

And trying to mix the two causes problems, because there's no guarantee the default email provider will be SMTP. (And if it is, users might not specify all the options.) Here's a test that fails even with your latest updates:

```python
# tests/mail/tests.py
class MoreSettingsTests(SimpleTestCase):
    @override_settings(
        EMAIL_PROVIDERS={
            "default": {
                "BACKEND": "third.party.EmailBackend",
            },
            "alternative": {
                "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
                "OPTIONS": {"host": "smtp.example.com"},
            },
        }
    )
    def test_smtp_provider_not_default(self):
        backend = mail.get_connection(provider="alternative")
        self.assertEqual(backend.port, 25)  # should be the default port
        # AssertionError: None != 25
```

Another way to think about this: the old `EMAIL_PORT` setting "belonged to" the SMTP EmailBackend, which could count on it being there. But `EMAIL_PROVIDERS["default"]["OPTIONS"]` *doesn't* belong to the SMTP EmailBackend, or to *any* particular backend. So any code that assumes `EMAIL_PROVIDERS["default"]` is an *SMTP* backend or has particular OPTIONS is going to be a problem.

**So, where should the SMTP EmailBackend now look for its defaults?**

Since there are no longer any SMTP-specific EMAIL_ settings, I think the SMTP defaults must move into the SMTP backend itself. You just did this for the "check deprecated settings" case; we also need it in the "initialized from EMAIL_PROVIDERS" case:

```python
# django/core/mail/backends/smtp.py:
class EmailBackend(BaseEmailBackend):
    DEFAULT_HOST = "localhost"
    DEFAULT_PORT = 25

    def __init__(self, ..., host=None, port=None, ..., provider=None):
        ...
        if provider is not None:
            # Being initialized from EMAIL_PROVIDERS.
            self.host = host or self.DEFAULT_HOST
            self.port = port or self.DEFAULT_PORT
            ...
        else:
            # RemovedInDjango70Warning: Check deprecated settings...
            self.host = host or getattr(settings, "EMAIL_HOST", self.DEFAULT_HOST)
            self.port = port or getattr(settings, "EMAIL_PORT", self.DEFAULT_PORT)
            ...
        ...
```

Once you've done that, several things get simpler:

* You can remove the OPTIONS dict from the global EMAIL_PROVIDERS default:
    ```python
    # django/conf/global_settings.py:
    EMAIL_PROVIDERS = {
        "default": {
            "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
            # No need for OPTIONS here: SMTP defaults are built into the SMTP backend.
        },
    }
    ```

* You can remove the special case handling for deprecated EMAIL_ settings in `LazySettings.__getattr__()`. (The SMTP backend properly checks the deprecated settings where necessary for compatibility.)

* In the tests, there's no longer a need to force an options dict into backend constructors. You can change `backend = smtp.EmailBackend(**settings.EMAIL_PROVIDERS["default"]["OPTIONS"])` back to its original `backend = smtp.EmailBackend()`. (Or `smtp.EmailBackend(provider="anything")` to send it down the non-deprecated path.) Similarly, you should be able to remove most of the cases where you're copying `settings.EMAIL_PROVIDERS["default"]["OPTIONS"]` in `@override_settings`.

Let me know if that makes sense, or if I've missed something.

medmunds commented:
I wrote:
> * In the tests … You can change `backend = smtp.EmailBackend(**settings.EMAIL_PROVIDERS["default"]["OPTIONS"])` back to its original `backend = smtp.EmailBackend()`. (Or `smtp.EmailBackend(provider="anything")` to send it down the non-deprecated path.)

Although that's true, I'm thinking there's a better approach. If the purpose of the test is to check that settings work correctly, the test really should be using `mail.get_connection()` to instantiate the backend. Django has always documented that get_connection() is the way to get "an instance of the email backend that you can use." Constructing a backend directly happened to work with the deprecated settings (and still does), but for settings-related tests we should probably be using the official API.

```patch
    @override_settings(
-       EMAIL_SSL_KEYFILE="foo"
+       EMAIL_PROVIDERS={
+           "default": {
+               "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
+               "OPTIONS": {"ssl_keyfile": "foo"},
+           },
+       }
    )
    def test_email_ssl_keyfile_use_settings(self):
-       backend = smtp.EmailBackend()
+       backend = mail.get_connection()
        self.assertEqual(backend.ssl_keyfile, "foo")
```

(Maybe we could figure out some sort of test helper that would override the OPTIONS for the default provider, to keep the tests more concise.)

medmunds commented on tests/mail/tests.py:
```diff
@@ -1605,12 +1612,220 @@ def test_8bit_non_latin(self):
+            ):
+                pass
+
+    def test_third_party_app_using_deprecated_default_configuration(self):
+        self.assertEqual(
+            settings.EMAIL_BACKEND,
+            "django.core.mail.backends.locmem.EmailBackend",
+        )
+        self.assertEqual(settings.EMAIL_HOST, "localhost")
+        self.assertEqual(settings.EMAIL_PORT, 25)
```
So, the issue with third party wrapper backends is not that they try to read these deprecated settings directly, but that they try to construct *other* backends by calling `get_connection(backend)` without a provider. And those *other* backends (like smtp.EmailBackend) still need to work properly with their own deprecated settings:

```python
    @override_settings(
        EMAIL_BACKEND="third.party.WrapperEmailBackend",
        THIRD_PARTY_WRAPPED_BACKEND="django.core.mail.backends.smtp.EmailBackend",
        EMAIL_HOST="smtp.example.com",
    )
    def test_third_party_wrapper_backend(self):
        # A wrapper backend does something like this:
        wrapped = mail.get_connection(settings.THIRD_PARTY_WRAPPED_BACKEND)
        # And it expects its wrapped connection to be initialized properly:
        self.assertIsInstance(wrapped, smtp.EmailBackend)
        self.assertEqual(wrapped.host, "smtp.example.com")  # from settings
        self.assertEqual(wrapped.port, 25)  # from defaults
```

Ditto for the similar tests below.

medmunds commented on tests/mail/tests.py:
```diff
@@ -1605,12 +1612,220 @@ def test_8bit_non_latin(self):
+        self.assertEqual(settings.EMAIL_HOST, "localhost")
+        self.assertEqual(settings.EMAIL_PORT, 25)
+
+    @override_settings(
+        EMAIL_PROVIDERS={
+            "default": {
+                "BACKEND": "django.core.mail.backends.dummy.EmailBackend",
+            },
+        },
+    )
+    def test_third_party_app_using_deprecated_backend_configuration(self):
+        self.assertEqual(
+            settings.EMAIL_BACKEND,
+            "django.core.mail.backends.dummy.EmailBackend",
+        )
```
*Maybe* EMAIL_BACKEND is the exception to my earlier comment. But I can't think of any third party apps that would rely on settings.EMAIL_BACKEND being available like this. (It shouldn't be necessary for wrapper backends, and other apps would normally let get_connection() handle EMAIL_BACKEND.) 

Did you have some other example?

jrief replied:
This is just to test if all of the new configuration settings are passed as deprecated EMAIL_… settings to external 3rd party pass. And `EMAIL_BACKEND` is one of them.

medmunds commented on tests/mail/tests.py:
```diff
@@ -1605,12 +1612,220 @@ def test_8bit_non_latin(self):
+        EMAIL_BACKEND="django.core.mail.backends.filebased.EmailBackend",
+        EMAIL_FILE_PATH="/tmp/foo",
+    )
+    def test_deprecated_filebased_settings_with_get_connection_kwargs(self):
+        # kwarg params to get_connection() override settings
+        connection = mail.get_connection(file_path="/tmp/bar")
+        self.assertIsInstance(connection, filebased.EmailBackend)
+        self.assertPathEqual(connection.file_path, "/tmp/bar")  # from kwargs
+
+    @override_settings(
+        EMAIL_BACKEND="django.core.mail.backends.dummy.EmailBackend",
+        EMAIL_FILE_PATH="/tmp/foo",
+    )
+    def test_get_connection_specified_filebased_backend(self):
+        # specific backend overrides EMAIL_BACKEND setting; kwargs work as above
+        connection = mail.get_connection(
+            "django.core.mail.backends.filebased.EmailBackend",
+            file_path="/tmp/bar",
+        )
+        self.assertIsInstance(connection, filebased.EmailBackend)
+        self.assertPathEqual(connection.file_path, "/tmp/bar")  # from kwargs
+
+
 class BaseEmailBackendTests(MailTestsMixin):
     email_backend = None
 
     @classmethod
     def setUpClass(cls):
-        cls.enterClassContext(override_settings(EMAIL_BACKEND=cls.email_backend))
+        cls.enterClassContext(
+            override_settings(
+                EMAIL_PROVIDERS={
+                    "default": dict(
+                        settings.EMAIL_PROVIDERS["default"],
+                        BACKEND=cls.email_backend,
+                    ),
+                }
+            )
+        )
         super().setUpClass()
```
I think when these tests were written, `@override_settings` couldn't be used as a class decorator.

Now that it can, it would be much less confusing (IMO) to remove this base setUpClass() method (and the `email_backend` properties), and instead decorate each backend-specific test class with its own settings:

```patch
+ @override_settings(
+     EMAIL_PROVIDERS={
+         "default": {
+             "BACKEND": "django.core.mail.backends.locmem.EmailBackend",
+         },
+     },
+ )
  class LocmemBackendTests(BaseEmailBackendTests, SimpleTestCase):
-     email_backend = "django.core.mail.backends.locmem.EmailBackend"
      ...

+ @override_settings(
+     EMAIL_PROVIDERS={
+         "default": {
+             "BACKEND": "django.core.mail.backends.console.EmailBackend",
+         },
+     },
+ )
  class ConsoleBackendTests(BaseEmailBackendTests, SimpleTestCase):
-     email_backend = "django.core.mail.backends.console.EmailBackend"
      ...
```

(You won't be able to use `@override_settings` like this for SMTPBackendTestsBase, but it already has its own setUpClass() that makes this one in base redundant.)

jrief replied:
I just don't want to change that much of the existing testing logik here. There is always the risk from a fellow asking to separate this into separate pull requests.

medmunds commented on tests/mail/tests.py:
```diff
@@ -2145,8 +2367,16 @@ def setUpClass(cls):
         cls._settings_override = override_settings(
-            EMAIL_HOST=cls.smtp_controller.hostname,
-            EMAIL_PORT=cls.smtp_controller.port,
+            EMAIL_PROVIDERS={
+                "default": {
+                    "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
+                    "OPTIONS": dict(
+                        settings.EMAIL_PROVIDERS["default"]["OPTIONS"],
+                        host=cls.smtp_controller.hostname,
+                        port=cls.smtp_controller.port,
+                    ),
+                },
+            }
```
Copying in the default OPTIONS should be unnecessary here. (If it *is* somehow necessary, that invalidates these tests, because users won't be copying in the default OPTIONS in their own settings.)

```suggestion
            EMAIL_PROVIDERS={
                "default": {
                    "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
                    "OPTIONS": dict(
                        host=cls.smtp_controller.hostname,
                        port=cls.smtp_controller.port,
                    ),
                },
            }
```

medmunds commented on tests/mail/tests.py:
```diff
@@ -2177,26 +2407,47 @@ def get_smtp_envelopes(self):
     @override_settings(
-        EMAIL_HOST_USER="not empty username",
-        EMAIL_HOST_PASSWORD="not empty password",
+        EMAIL_PROVIDERS={
+            "default": {
+                "OPTIONS": dict(
+                    settings.EMAIL_PROVIDERS["default"]["OPTIONS"],
+                    username="not empty username",
+                    password="not empty password",
+                ),
+            },
+        }
     )
     def test_email_authentication_use_settings(self):
-        backend = smtp.EmailBackend()
+        backend = smtp.EmailBackend(**settings.EMAIL_PROVIDERS["default"]["OPTIONS"])
         self.assertEqual(backend.username, "not empty username")
         self.assertEqual(backend.password, "not empty password")
```
See my earlier [comment about settings tests](https://github.com/django/django/pull/18421/#issuecomment-2623482172).

Also, if you wanted to cut down on the duplicated settings dicts, you could combine this test_email_authentication_use_settings() and the following test_email_authentication_override_settings() in a single function with a shared `@override_settings`, using subTest() to separate the two cases:

```python
    @override_settings(
        EMAIL_PROVIDERS={
            "default": {
                "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
                "OPTIONS": {
                    "username": "not empty username",
                    "password": "not empty password",
                },
            },
        }
    )
    def test_email_authentication_options(self):
        with self.subTest("use settings"):
            backend = mail.get_connection()
            self.assertEqual(backend.username, "not empty username")
            self.assertEqual(backend.password, "not empty password")

        with self.subTest("override settings"):
            backend = mail.get_connection(username="username", password="password")
            self.assertEqual(backend.username, "username")
            self.assertEqual(backend.password, "password")
```

(and similar for the large number of settings-related tests that follow)

medmunds commented on tests/mail/tests.py:
```diff
@@ -2441,7 +2832,7 @@ def test_encodes_idna_in_smtp_commands(self):
             },
         )
-        backend = smtp.EmailBackend()
+        backend = smtp.EmailBackend(**settings.EMAIL_PROVIDERS["default"]["OPTIONS"])
```
```suggestion
        backend = mail.get_connection()
```

(And similar in other tests.)

medmunds commented on tests/mail/tests.py:
```diff
@@ -2334,9 +2689,45 @@ def __init__(self, *args, **kwargs):
+        }
+    )
+    def test_email_alternative_provider(self):
+        backend = smtp.EmailBackend(provider="alternative")
```
```suggestion
        backend = mail.get_connection(provider="alternative")
```

medmunds reviewed (changes requested):
(Feedback on tests and a few other areas.)

medmunds commented:
> @medmunds all very valid annotations you pointed out here. Today I found some time to work on it again.

@jrief great! I'll wait to go through it again until you're ready—just click the "Re-request review" button by my name in the Reviewers panel.

Before that, it would be helpful if you could go through the open review comments and either click "Resolved" if you've addressed it, or add a reply if further discussion is needed. (I'm not able to resolve my own comments since it's your PR. Also, remember that GitHub has a bad habit of hiding unresolved conversations behind a "load more" link.)

jrief commented:
@medmunds a question about the deprecation warnings. In the comments and release notes I now changed them to version 6.1. This means that starting from version 6.0 a user would get a warning that for instance `settings.EMAIL_BACKEND` is deprecated in version 6.1. It then will be removed altogether in version 7.0. It that correct?

Btw. who is going to remove those settings and the associated unit tests then in the future?

medmunds commented:
> a question about the deprecation warnings. In the comments and release notes I now changed them to version 6.1. This means that starting from version 6.0 a user would get a warning that for instance `settings.EMAIL_BACKEND` is deprecated in version 6.1. It then will be removed altogether in version 7.0. It that correct?

Very close. Yes, the old behavior will be removed altogether in version 7.0. But it is considered deprecated *immediately* in version *6.0,* not 6.1. (The warning message only says when the feature will be removed, not when it was deprecated.)

* This PR will be released in Django 6.0 (assuming it is completed and merged before the [6.0 alpha](https://code.djangoproject.com/wiki/Version6.0Roadmap) on ~2025-09-17).
* The deprecation period for Django 6.0 lasts until Django 7.0 (the next major release). So any behavior deprecated in this PR is "deprecated in Django 6.0" and will be "removed in Django 7.0."
*  During the deprecation period, old behavior must continue to work but issue a deprecation warning. In your code, you should use `RemovedInDjango70Warning` everywhere. (Not `RemovedInDjango61Warning`. Those might be left over from earlier versions of this PR that were aiming for the Django 5.2 release.)
* In docs/releases/6.0.txt, you would describe the necessary changes in the "Features deprecated in 6.0" section. (You might also need to briefly list them in docs/internals/deprecation.txt under "removed in 7.0." I'm not entirely clear on the overlap between those two pages.)
* Elsewhere in the docs, you would use `.. deprecated:: 6.0` to identify old behavior and explain the replacement.

See [*Deprecating a feature*](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature) in the contributing docs for more info and examples.

> Btw. who is going to remove those settings and the associated unit tests then in the future?

A Django fellow will do that as part of preparing the codebase for Django 7.1. They just search for `RemovedInDjango70Warning`. That's why it's important to include a RemovedInDjango70Warning *comment* near any code that doesn't issue a warning but still needs to be removed or changed.

medmunds commented:
@jrief one other thing: there's no such thing as RemovedInDjango62Warning. That had confused me too: it jumps from RemovedInDjango61Warning directly to RemovedInDjango70Warning. It turns out that's because Django's deprecation period lasts until the next major release *or* at least two minor releases, whichever is longer.

jrief commented:
@medmunds hopefully now everything is fixed.

I added a comment `# RemovedInDjango70Warning` to all blocks which must be removed for Django-7.0.

I added a short section on the new configuration directive in `docs/releases/6.0.txt` and `docs/internals/deprecation.txt`.

jrief commented:
@medmunds in your opinion, is this pull request now ready to be reviewed by the fellows?

medmunds commented:
> @medmunds in your opinion, is this pull request now ready to be reviewed by the fellows?

@jrief there are a couple of design-level questions that need to be resolved. I'm going to leave a new review focusing solely on those.

Once we figure them out, I *think* there are a handful of other things the fellows might want cleaned up (e.g., in the docs and testing deprecated behavior). It'll be easier to look at those once the big design questions are settled.

Also, since I wasn't present for the design discussion at PyCon (and haven't seen any notes from it), it's quite possible I'm misunderstanding some things. If there was anyone in that discussion who could be helpful here, I think it would be fine to @ them in.

medmunds commented on django/conf/__init__.py:
```diff
@@ -74,7 +117,23 @@ def __getattr__(self, name):
+        except AttributeError:
+            # Special case for EMAIL_* settings. To avoid breaking existing third-party
+            # apps accessing the current Django settings. RemovedInDjango70Warning.
+            if name == "EMAIL_BACKEND":
+                return self.EMAIL_PROVIDERS["default"]["BACKEND"]
```
**If Q1a:** remove this code.

**If Q1b:**
- Replace hard-coded "default" with DEFAULT_EMAIL_PROVIDER_ALIAS.
- Needs a deprecation warning. This warns code that tries to *use* the deprecated setting. (It's the counterpart to your existing warning to users who *define* EMAIL_BACKEND in settings.py.)

```suggestion
            if name == "EMAIL_BACKEND":
                warnings.warn(RemovedInDjango70Warning, "...")
                return self.EMAIL_PROVIDERS[self.DEFAULT_EMAIL_PROVIDER_ALIAS]["BACKEND"]
```

The warning message is maybe something like, "Using settings.EMAIL_BACKEND is deprecated. Redesign code to work with EMAIL_PROVIDERS." 

(I don't think there's a specific replacement to suggest. The warning should *not* recommend switching to `settings.EMAIL_PROVIDERS["default"]["BACKEND"]`. Code that is reading EMAIL_BACKEND will need to be re-thought for a world with multiple providers.)

jrief replied:
I would opt for Q1b. Today I attempted to implement this and add a warining whenever code accesses `settings.EMAIL_BACKEND`. However, I still get some failing unit tests in code not associated with the maling subsystem. Please have a look at my recent changes.

For Q2, I still have to evaluate the options.

medmunds replied:
For both Q1 and Q2, the "a" option is "all or nothing" approach discussed in the recent ticket comments, and the "b" option is various attempts to allow EMAIL_PROVIDERS and deprecated EMAIL_ settings to coexist in the same ~~user settings file~~ project.

In the ticket, you said you thought "all or nothing" was the only viable approach. (And I tend to agree.) Has something changed your position?

medmunds commented on django/conf/__init__.py:
```diff
@@ -74,7 +117,23 @@ def __getattr__(self, name):
+            # apps accessing the current Django settings. RemovedInDjango70Warning.
+            if name == "EMAIL_BACKEND":
+                return self.EMAIL_PROVIDERS["default"]["BACKEND"]
+            if name in DEPRECATED_EMAIL_SETTINGS:
+                backend = import_string(getattr(self, "EMAIL_BACKEND"))(
+                    provider="default"
+                )
+                if name == "EMAIL_HOST_USER":
+                    return backend.username
+                if name == "EMAIL_HOST_PASSWORD":
+                    return backend.password
+                return getattr(backend, f"{name[6:].lower()}")
```
**Q2:** There is no way to make this work as intended. Remove it.

medmunds commented on django/conf/__init__.py:
```diff
@@ -180,6 +239,14 @@ def __init__(self, settings_module):
+        is_overridden_EMAIL_PROVIDERS = self.is_overridden("EMAIL_PROVIDERS")
+        for deprecated_setting in DEPRECATED_EMAIL_SETTINGS:
+            if self.is_overridden(deprecated_setting):
+                if is_overridden_EMAIL_PROVIDERS:
+                    raise EmailImproperlyConfigured(deprecated_setting)
```
**If Q2b:** remove the error here. (Maybe add a warning about potential confusion between EMAIL_PROVIDERS and the deprecated setting?)

medmunds commented on django/conf/__init__.py:
```diff
@@ -224,6 +291,17 @@ def __getattr__(self, name):
         self._deleted.discard(name)
+
+        # RemovedInDjango70Warning.
+        if name == "EMAIL_PROVIDERS" and any(
+            self.is_overridden(f) for f in DEPRECATED_EMAIL_SETTINGS
+        ):
+            raise EmailImproperlyConfigured(name)
+        if name in DEPRECATED_EMAIL_SETTINGS:
+            if self.is_overridden("EMAIL_PROVIDERS"):
+                raise EmailImproperlyConfigured(name)
+            warn_about_deprecated_email_setting(name)
```
**If Q2b:** Remove the errors here. (Maybe add a warning about potential confusion between EMAIL_PROVIDERS and the deprecated setting?)

medmunds commented on django/conf/global_settings.py:
```diff
@@ -186,30 +186,17 @@ def gettext_noop(s):
-
-# Port for sending email.
-EMAIL_PORT = 25
+# The email provider settings. If left empty, will default to the SMTP backend.
+# For possible shortcuts see django.core.mail.
+EMAIL_PROVIDERS = {
+    "default": {
+        "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
+    },
+}
```
Need to define (and use) global default for DEFAULT_EMAIL_PROVIDER_ALIAS.

```suggestion
# The EMAIL_PROVIDERS key to use for sending if no provider is specified.
DEFAULT_EMAIL_PROVIDER_ALIAS = "default"

# The email provider settings. If left empty, will default to the SMTP backend.
# For possible shortcuts see django.core.mail.
EMAIL_PROVIDERS = {
    DEFAULT_EMAIL_PROVIDER_ALIAS: {
        "BACKEND": "django.core.mail.backends.smtp.EmailBackend",
    },
}
```

medmunds commented on django/conf/global_settings.py:
```diff
@@ -186,30 +186,17 @@ def gettext_noop(s):
 # Whether to send SMTP 'Date' header in the local time zone or in UTC.
 EMAIL_USE_LOCALTIME = False
 
-# Optional SMTP authentication information for EMAIL_HOST.
-EMAIL_HOST_USER = ""
-EMAIL_HOST_PASSWORD = ""
-EMAIL_USE_TLS = False
-EMAIL_USE_SSL = False
-EMAIL_SSL_CERTFILE = None
-EMAIL_SSL_KEYFILE = None
-EMAIL_TIMEOUT = None
```
**If Q2b:** we would need to keep defaults for these deprecated settings (plus EMAIL_HOST and EMAIL_PORT above).

medmunds commented on django/core/mail/__init__.py:
```diff
@@ -44,16 +45,47 @@
+
+    options = {}
+
+    # RemovedInDjango70Warning: Use deprecated EMAIL_BACKEND if set.
+    if not backend and not provider:
+        backend = getattr(settings, "EMAIL_BACKEND", None)
```
**If Q1b:** There's a bad interaction between the LazySettings EMAIL_BACKEND fallback and this code. To fix:

```suggestion
    # RemovedInDjango70Warning: Use deprecated EMAIL_BACKEND if specifically set.
    if not backend and not provider and settings.is_overridden("EMAIL_BACKEND"):
        backend = settings.EMAIL_BACKEND
```

Once you've made this change, there's no longer any need for the EmailBackend implementations to be looking directly at `EMAIL_PROVIDERS[provider]["OPTIONS"]`. (Which they shouldn't be doing.)

**If Q1a:** the code here is correct as written. (Because you'd be removing the LazySettings EMAIL_BACKEND fallback.)

medmunds commented on django/core/mail/backends/filebased.py:
```diff
@@ -9,13 +9,24 @@
-        if file_path is not None:
-            self.file_path = file_path
+
+        if provider is not None:
+            # Being initialized from EMAIL_PROVIDERS.
+            options = settings.EMAIL_PROVIDERS[provider].get("OPTIONS", {})
+            file_path = options.get("file_path")
```
An EmailBackend **must not** examine settings.EMAIL_PROVIDERS directly. (That is mail.get_connection()'s responsibility.)

```suggestion
        if provider is not None:
            # Being initialized from EMAIL_PROVIDERS.
            pass
            # file_path arg is already set by mail.get_connection(). Do not override it here.
            # (If you want to raise ImproperlyConfigured for missing file_path, do that here.)
```

medmunds commented on django/core/mail/backends/smtp.py:
```diff
@@ -28,25 +31,66 @@ def __init__(
-            settings.EMAIL_SSL_CERTFILE if ssl_certfile is None else ssl_certfile
-        )
+
+        if provider is not None:
+            # Being initialized from EMAIL_PROVIDERS.
+            options = settings.EMAIL_PROVIDERS[provider].get("OPTIONS", {})
+            self.host = host or options.get("host", self.DEFAULT_HOST)
+            self.port = port or options.get("port", self.DEFAULT_PORT)
+            self.username = options.get("username") if username is None else username
+            self.password = options.get("password") if password is None else password
+            self.use_tls = options.get("use_tls") if use_tls is None else use_tls
+            self.use_ssl = options.get("use_ssl") if use_ssl is None else use_ssl
+            self.ssl_keyfile = (
+                options.get("ssl_keyfile") if ssl_keyfile is None else ssl_keyfile
+            )
+            self.ssl_certfile = (
+                options.get("ssl_certfile") if ssl_certfile is None else ssl_certfile
+            )
+            self.timeout = options.get("timeout") if timeout is None else timeout
```
An EmailBackend **must not** examine settings.EMAIL_PROVIDERS directly. (That is mail.get_connection()'s responsibility.)

```suggestion
        if provider is not None:
            # Being initialized from EMAIL_PROVIDERS.
            self.host = self.DEFAULT_HOST if host is None else host
            self.port = self.DEFAULT_PORT if port is None else port
            self.username = username
            self.password = password
            self.use_tls = False if (use_tls is None) else use_tls
            self.use_ssl = False if (use_ssl is None) else use_ssl
            self.ssl_keyfile = ssl_keyfile
            self.ssl_certfile = ssl_certfile
            self.timeout = timeout
```

medmunds reviewed (changes requested):
@jrief There are a couple of design-decision questions we need to resolve before this can move forward.

(I'm going to refer back to these questions and options for solutions in my inline review comments: Q1, Q2b, etc.)

**Q1: Should `settings.EMAIL_BACKEND` be readable from EMAIL_PROVIDERS?**

If settings.py defines EMAIL_PROVIDERS but not deprecated EMAIL_BACKEND, what happens with code that tries to use `django.conf.settings.EMAIL_BACKEND`?

There are two reasonable options here:

Q1a. It doesn't work: if EMAIL_BACKEND *isn't* defined in settings.py, code that tries to read `settings.EMAIL_BACKEND` gets an AttributeError. That means users cannot change their settings.py to EMAIL_PROVIDERS until they update *all* their code that tries to read `settings.EMAIL_BACKEND` to be compatible with EMAIL_PROVIDERS. And we document this.

Q1b. It works with a deprecation warning: reading `settings.EMAIL_BACKEND` provides the value of the default EMAIL_PROVIDER's BACKEND. You've implemented that in `LazySettings.__getattr__`. If we want to stick with this approach, there are a couple of things to fix (see review comments in the code).

To decide, we need to know whether this is an important case to support. You've added a specific test for this behavior. Is it based on known third-party code? Or do you have an example scenario where this seems likely? (That is, the user must change settings.py to EMAIL_PROVIDERS but also keep using third-party code that reads `settings.EMAIL_BACKEND`?)

I think option Q1a is much simpler, both to implement and to explain. But Q1b is possible if there's a good reason for it.


**Q2: Should `settings.EMAIL_HOST` and other deprecated EMAIL settings be readable from EMAIL_PROVIDERS?**

This is similar to the first question, but for all the other deprecated EMAIL settings.

You've tried to implement reading these in `LazySettings.__getattr__`, by instantiating the default EMAIL_PROVIDER's backend and accessing the corresponding attribute—e.g., reading `settings.EMAIL_USER` from `backend.username`.

But that code assumes the default backend is Django's smtp.EmailBackend. It simply won't work in general. E.g., what happens when third-party code tries to read `settings.EMAIL_HOST` if the user has these settings:

```python
EMAIL_PROVIDERS = {
    "default": {
        "BACKEND": "some.non_django.EmailBackend",
        "OPTIONS": {
            "api_key": "123",
        },
    },
}
```

I don't think there's any way to make LazySettings work as you intend. So we should consider alternatives. I see two options:

Q2a. It doesn't work: if EMAIL_HOST *isn't* defined in settings.py, code that tries to read `settings.EMAIL_HOST` gets an AttributeError. That means users cannot change their settings.py to EMAIL_PROVIDERS until they update *all* code that tries to read `settings.EMAIL_HOST` and other deprecated settings to be compatible with EMAIL_PROVIDERS. And we document this.

Q2b. We allow settings.py to define both EMAIL_PROVIDERS and deprecated EMAIL_HOST at the same time. So code that reads `settings.EMAIL_HOST` works (but should get a deprecation warning). Users are responsible for keeping their deprecated EMAIL settings in sync with EMAIL_PROVIDERS until *all* their code that reads deprecated settings has been updated to use EMAIL_PROVIDERS instead.

Again, is this a case we really need to handle? I'm not aware of third party code that borrows Django's EMAIL settings like this, but you may have some example in mind?

If not, Q2a is much simpler. If so, there's probably a way to make Q2b work, but there's a high potential for user confusion if their settings get out of sync.

medmunds commented:
@jrief GitHub has decided to hide a critical inline comment from my review ([this one](#discussion_r1960274072)) 😠 . Be sure to "load more" hidden conversations.

medmunds commented:
> a couple of design-decision questions we need to resolve

After poking around in GitHub code search, my votes are Q1a and Q2a. That is, with a settings.py that defines EMAIL_PROVIDERS but not any deprecated EMAIL\_* settings, we would expect:

```pycon
>>> from django.conf import settings
>>> settings.EMAIL_BACKEND  # (error per Q1a)
AttributeError: 'Settings' object has no attribute 'EMAIL_BACKEND'
>>> settings.EMAIL_HOST_USER  # (error per Q2a)
AttributeError: 'Settings' object has no attribute 'EMAIL_HOST_USER'
```

The code searches find plenty of uses of the deprecated settings. But we only need to consider cases that appear to be libraries—where library *users* would be prevented from using EMAIL_PROVIDERS (at all) until the libraries stop using the deprecated settings. (Complete server implementations and other non-library projects—which include their own settings.py—can fix their own usage of the deprecated settings at the same time they change to EMAIL_PROVIDERS.)

> **Q1: Should `settings.EMAIL_BACKEND` be readable from EMAIL_PROVIDERS?**

[settings.EMAIL_BACKEND code search](https://github.com/search?q=%2F%28%3F-i%29%5Cbsettings.EMAIL_BACKEND%5Cb%2F+content%3Adjango.conf+language%3Apython+NOT+is%3Afork+NOT+repo%3Adjango%2Fdjango&type=code): I found two examples of *library* code that reads settings.EMAIL_BACKEND:

* [django-mailer](https://github.com/pinax/django-mailer/blob/e743aa972c3fd58f6cf6c571e17fd88c86879419/src/mailer/engine.py#L176-L182): checks for EMAIL_BACKEND configuration loop
* [django-uwsgi](https://github.com/unbit/django-uwsgi/blob/f6d452427eedf499e167ef63dacae955b17d0949/django_uwsgi/management/commands/runuwsgi.py#L101-L103): checks if its custom EmailBackend is in use

These seem like reasonable uses, and their existence *slightly* tips the scales toward Q1b (LazySettings magic) for me. But not enough to justify the additional complexity.

Put another way: the response to, "I'm using django-mailer and get an AttributeError when I try to configure it using EMAIL_PROVIDERS," should be, "Then keep using EMAIL_BACKEND as shown in django-mailer's docs, until django-mailer is updated to work with EMAIL_PROVIDERS."


> **Q2: Should `settings.EMAIL_HOST` and other deprecated EMAIL settings be readable from EMAIL_PROVIDERS?**

[settings.EMAIL_&lt;others&gt; code search](https://github.com/search?q=%2F%28%3F-i%29%5Cbsettings%5C.EMAIL_%28HOST%7CPORT%7CHOST_USER%7CHOST_PASSWORD%7CUSE_TLS%7CUSE_SSL%7CTIMEOUT%7CSSL_KEYFILE%7CSSL_CERTFILE%29%5Cb%2F+content%3Adjango.conf+language%3Apython+NOT+is%3Afork+NOT+path%3Asettings.py&type=code): I found only one example of *library* code that reads other deprecated settings:

* [healthchecks](https://github.com/healthchecks/healthchecks/blob/0e33dc5425214abfcebf74f70fdb49de0319df60/hc/api/apps.py#L45-L52): issues a system checks warning if settings.EMAIL_HOST is falsey

I didn't see *any* libraries trying to read the other deprecated settings.[^1]

[^1]: But there's a *lot* of non-library code that sends email with `from_email=settings.EMAIL_HOST_USER`, which seems like a misunderstanding of DEFAULT_FROM_EMAIL.

Same argument as above… Complaint: "I'm getting an AttributeError in healthchecks when I remove EMAIL_HOST from my settings.py." Solution: "Then don't remove EMAIL_HOST until healthchecks is updated. That means you'll need to wait to switch to EMAIL_PROVIDERS, too."

[Standard code search disclaimer: these are illustrative examples, *not* a complete list of projects that might be affected. Not all projects are in GitHub; not all GitHub projects are in public code search; my searches miss things like `getattr(settings, "EMAIL_BACKEND")`; etc.]

medmunds commented:
@jrief sorry, there is one other case to consider. Imagine a user has just upgraded to Django 6.0 and has not modified their settings file at all yet. (So they don't have EMAIL_PROVIDERS defined in their settings.) In that case, *any* user or library code that tries to access any deprecated setting must continue to work, but should get a deprecation warning.

Once the user opts in to using EMAIL_PROVIDERS, the compatibility guarantee no longer applies. If EMAIL_PROVIDERS is defined in settings.py, then:
* it's a configuration error to also define other deprecated EMAIL settings (as you've already implemented)
* it's an AttributeError to try to read the deprecated settings (my options "Q1a" and "Q2a" in the earlier comment)

But if EMAIL_PROVIDERS is *not* defined, we need to make all the deprecated settings readable with their original default values. I think the `dep_SETTING_NAME = ...` pattern you'd seen used in global_settings.py is part of implementing that. Here's a possible approach:

```python
# django/conf/__init__.py

class LazySettings(LazyObject):
    def __getattr__(self, name):
        ...
        try:
            val = getattr(_wrapped, name)
        except AttributeError:
            # RemovedInDjango70Warning: If EMAIL_PROVIDERS is not being used,
            # provide original default values for deprecated EMAIL settings.
            if name in DEPRECATED_EMAIL_SETTINGS and not self.is_overridden(
                "EMAIL_PROVIDERS"
            ):
                self._show_deprecation_warning(
                    f"Using the {name} setting is deprecated."
                    " Update code to work with EMAIL_PROVIDERS.",
                    RemovedInDjango70Warning,
                )
                val = getattr(_wrapped, f"dep_{name}")
            else:
                raise
        ...
```

```python
# django/conf/global_settings.py

# RemovedInDjango70Warning: Default values for deprecated EMAIL settings.
# (Used in LazySettings.__getattr__.)
dep_EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
dep_EMAIL_HOST = "localhost"
dep_EMAIL_PORT = 25
dep_EMAIL_HOST_USER = ""
dep_EMAIL_HOST_PASSWORD = ""
dep_EMAIL_USE_TLS = False
dep_EMAIL_USE_SSL = False
dep_EMAIL_SSL_CERTFILE = None
dep_EMAIL_SSL_KEYFILE = None
dep_EMAIL_TIMEOUT = None
```

medmunds commented on tests/mail/tests.py:
```diff
@@ -2334,11 +2693,50 @@ def __init__(self, *args, **kwargs):
         self.assertEqual(myemailbackend.connection.timeout, 42)
```
test_connection_timeout_custom() is invalid, because it directly instantiates a (subclass of) smtp.EmailBackend rather than using mail.get_connection():

```patch
  def test_connection_timeout_custom(self):
      ...
-     myemailbackend = MyEmailBackend()
+     myemailbackend = mail.get_connection()
```

To make this test work, you'll need to make the custom MyEmailBackend importable and then refer to its import path in EMAIL_PROVIDERS. There are a several ways to do that:
* The cleanest is probably to move it into to custombackend.py and rename it something like CustomTimeoutBackend. Then override EMAIL_PROVIDERS to set BACKEND to `"mail.custombackend.CustomTimeoutBackend"`. 
* Or if you want to keep MyEmailBackend defined inside this test case, then temporarily install it in the current module: `sys.modules[self.__module__].MyEmailBackend = MyEmailBackend` and override EMAIL_PROVIDERS to set BACKEND to `f"{self.__module__}.MyEmailBackend"`. (And delattr it out of the module when you're done, in a try-finally block. The other option seems cleaner.)

Please *don't* try to make this invalid test work by reading EMAIL_PROVIDERS directly inside the smtp EmailBackend implementation. (That's the responsibility of mail.get_connection().)

medmunds commented:
@jrief Is there anything I could do to help move this forward? For example, I might be able to open a PR on your PR (against your PR branch in your fork of Django), which you could review and merge into your PR or request that I make changes.

jrief commented:
@medmunds I'm so sorry for not proceeding with this. I just didn't have time anymore to work on it. If you can work on it, please feel free to do so. I usually finish a job that I have started, so whenever I have some extra time, I will proceed.

medmunds commented:
@jrief no apologies necessary—I completely understand. I'll try to get back into it next week and let you know if I make any progress.

jrief commented:
@medmunds Is there still anything to be fixed on this issue (except rebasing), to make this a pull request for review by the Django fellows?

medmunds commented:
@jrief It doesn't look like any of the comments from [my mid-Feb review](https://github.com/django/django/pull/18421#pullrequestreview-2624473570) have been addressed. Per the [follow-up comment](https://github.com/django/django/pull/18421#issuecomment-2667257997) I suggest the "a" option for each of the design decisions.

(Also, it seems GitHub has introduced a "new PR experience" that hides even more current comments than the old experience. You may need to disable that near the top of the Files tab.)

> I retested the concern you mentioned in your previous [comment](https://github.com/django/django/pull/18421/#issuecomment-2669699244). Actually, accessing `settings.EMAIL_HOST` and the other deprecated settings works in the current branch. I added an [extra test](https://github.com/jrief/django/blob/58e1aebfeb8c558b3237003aac10a809068b4f61/tests/mail/tests.py#L2426) to prove this.

The concern is about reading deprecated EMAIL_ settings where the user has *not* provided a settings.py value:

```python
    # No @override_settings()
    def test_deprecated_settings_have_default_values(self):
        # Django 6.0 default values for deprecated email settings:
        self.assertEqual(
            settings.EMAIL_BACKEND,
            "django.core.mail.backends.smtp.EmailBackend",
        )
        self.assertEqual(settings.EMAIL_HOST, "localhost")
        self.assertEqual(settings.EMAIL_PORT, 25)
        self.assertEqual(settings.EMAIL_HOST_USER, "")
        self.assertEqual(settings.EMAIL_HOST_PASSWORD, "")
        # Using assertIsFalse(...) rather than assertEqual(..., False)
        # (despite coding guidelines) to ensure exact Django 6.0 values.
        self.assertIsFalse(settings.EMAIL_USE_TLS)
        self.assertIsFalse(settings.EMAIL_USE_SSL)
        self.assertIsNone(settings.EMAIL_SSL_CERTFILE)
        self.assertIsNone(settings.EMAIL_SSL_KEYFILE)
        self.assertIsNone(settings.EMAIL_TIMEOUT)
```

medmunds commented:
@jrief I was going to take a crack at an updated version of this next week, if that's OK with you. (Don't want to step on any work you might have in progress.)

jacobtylerwalls commented:
I think he meant same release, separate PRs, with the merger "on their honor" to not merge stuff unless they feel sanguine (🩸) on getting all of it done by the freeze (🧊). To save the contributor having to rebase on so many moving targets.

nessita commented:
> I'd _probably_ handle this as a single PR with several commits

This is my strong preference too, so thank you for mentioning it. Well-contained commits could/would be, later on during the lifetime of the PR, cherry-picked into a separated PR for (as) early landing (as possible).
